pkg/capabilities/consensus/ocr3/capability.go

```diff
@@ -0,0 +1,172 @@
+package consensus
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+var info = capabilities.MustNewCapabilityInfo(
+	"ocr3",
+	capabilities.CapabilityTypeConsensus,
+	"OCR3 consensus exposed as a capability.",
+	"v1.0.0",
+)
+
+type capability struct {
+	services.StateMachine
+	capabilities.CapabilityInfo
+	store  *store
+	cancel func()
+
+	clock clockwork.Clock
+
+	newExpiryWorkerCh chan *request
+}
+
+func newCapability(s *store, clock clockwork.Clock) *capability {
+	o := &capability{
+		CapabilityInfo:    info,
+		store:             s,
+		newExpiryWorkerCh: make(chan *request),
+		clock:             clock,
+	}
+	return o
+}
+
+func (o *capability) Start(ctx context.Context) error {
+	return o.StartOnce("OCR3Capability", func() error {
+		ctx, c := context.WithCancel(ctx)
+		o.cancel = c
+		go o.loop(ctx)
+		return nil
+	})
+}
+
+func (o *capability) Stop(ctx context.Context) error {
+	return o.StopOnce("OCR3Capability", func() error {
+		o.cancel()
+		return nil
+	})
+}
+
+func (o *capability) RegisterToWorkflow(ctx context.Context, request capabilities.RegisterToWorkflowRequest) error {
+	return nil
+}
+
+func (o *capability) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+func (o *capability) Execute(ctx context.Context, callback chan<- capabilities.CapabilityResponse, request capabilities.CapabilityRequest) error {
+	// Receives and stores an observation to do consensus on
+	// Receives an aggregation method; at this point the method has been validated
+	// Returns the consensus result over a channel
+	r, err := o.unmarshalRequest(ctx, request, callback)
+	if err != nil {
+		return err
+	}
+
+	err = o.store.add(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	o.newExpiryWorkerCh <- r
+	return nil
+}
+
+func (o *capability) loop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case r := <-o.newExpiryWorkerCh:
+			go o.expiryWorker(ctx, r)
```

Comment:
consider time.AfterFunc()

Reply:
Hmmm... I tried using AfterFunc here but I don't think it works since we need to also listen to ctx.Done() in our case (to handle the capability being stopped).

Reply:
For that you can call Stop() on each object returned by AfteFunc
```
timer := time.AfterFunc
timer.Stop()
```

Reply:
Unless I'm missing something, I don't think that works: the docs for time.AfterFunc say that the call will block until the function has been called.

But in our case, we want to either exit without calling the function if its context is cancelled, or call the function.

I'm not able to account for that first case -- I think the only way to do that is a select on both the ctx.Done() and the timer's channel.

Reply:
(Also the channel for the returned timer is empty, so we can't do t := time.AfterFunc; and then <-t.C in a select)

Reply:
Hmm interesting. I rely on the Stop() call in another place so I need to check if this actually works as intended...
Feel free to go with you initial approach here.